rfcs/0183-versioned-flakes.md

```diff
@@ -0,0 +1,55 @@
+---
+feature: versioned-flakes
+start-date: 2024-11-25
+author: domenkozar
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+## Summary
+
+Introduce a standardized versioning schema for Nix flakes using [SemVer](https://semver.org/),
```

Comment:
There is a problem in semver: 

> how can we *prove* two packages are *semantically* same, even if their versions are different?.

Nix (as a package manager) think it is more reliable if: any package update will trigger   "full-rebuilds" for those dependent on it.

I just feel a little bit strange about introducing "SemVer" here.

Reply:
I'm confused, isn't this RFC about versioning *flakes*, not packages?

Reply:
> I'm confused, isn't this RFC about versioning flakes, not packages?

That's the interesting paradox: if we don't use semantic versioning (SemVer) for packages, why should we use SemVer for flakes?

If there are valid reasons for using SemVer with 'flake' dependencies, why wouldn't those same reasons apply to packages as well?

Reply:
How will you apply semver to nixpkgs packages?

Reply:
> That's the interesting paradox: if we don't use semantic versioning (SemVer) for packages, why should we use SemVer for flakes?
>
> If there are valid reasons for using SemVer with 'flake' dependencies, why wouldn't those same reasons apply to packages as well?

- Flake version is about the flake's API surface, while package version is about the package's API surface or whatever upstream wants to do, they are orthogonal to each other
- You can't and shouldn't force developers to use a certain version format for their software, so package versions should be as flexible as possible (nixpkgs of course has no restrictions at all really)
- Flake versions could be automatically resolved, it's well known where to search for versions and how to load them, while neither really applies to packages

I don't have an opinion on this RFC either way but the two situations are certainly very different.